tests/integration_tests/instances.py

```diff
@@ -134,6 +105,51 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             self.destroy()
 
 
+def install_proposed_image(instance):
```

Comment:
Moving these from methods on `IntegrationInstance` to functions which take an `IntegrationInstance` as the first argument feels a little weird to me. IIUC, it's because we want to be able to determine that the install method is valid in `setup_image` without launching the instance to find out.

I think this reveals a slight issue in the lifetimes of the objects we're dealing with: `pycloudlib.BaseInstance` instances are only created _after_ the underlying instance has been created (e.g. [`_BaseLXD` creates a `LXDInstance` after calling `lxd init`](https://github.com/canonical/pycloudlib/blob/7532404a5641a0f67ba6fa34898a6598b80d0816/pycloudlib/lxd/cloud.py#L289)). There is no way to reason about a `BaseInstance` instance without its corresponding cloud instance being created.

This is mirrored up into the cloud-init integration testing codebase: `IntegrationInstance`s constructor requires a `pycloudlib.BaseInstance` instance, so we can't reason about `IntegrationInstance`s without having launched them either.

This change would be much easier to make if `IntegrationInstance`s lifetime was _longer_ than that of its underlying cloud instance (and therefore its underlying `BaseInstance`): we would be able to reason about this without the launch.

This isn't to say that we should change that for this PR, but I am hesitant to further push this mismatch into our codebase by refactoring methods to functions.

What do you think to having a two-phase approach: one function which validates the install method (i.e. `get_install_method` but it doesn't return anything and so can be called whenever), and then `IntegrationInstance.install_new_cloud_init` determines which of the methods to use at that point in time?

(A potential improvement would be to have `get_install_method` return a value which `install_new_cloud_init` can select on: instead of `if source.startswith("ppa:"): self.install_ppa(repo=source)`, we could have `if install_method_we_just_looked_up == CloudInitSource.PPA: self.install_ppa(repo=source)`; this would avoid having to maintain the checks in two places.)